README.md
[![Build Status](https://travis-ci.org/Rob-Orb/picberry.svg?branch=master)](https://travis-ci.org/Rob-Orb/picberry)

# picberry
PIC Programmer using GPIO connector

Copyright 2017 Francesco Valla

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

## Overview

_picberry_ is a PIC programmer using GPIOs that doesn't require additional programming hardware.

It theorically supports dsPIC33E/PIC24E, dsPIC33F/PIC24H, PIC24FJ, PIC18FxxJxx, PIC32MX, PIC32MZ, PIC32MK and PIC16F150X families, but only some PICs have been tested.

### Supported hosts

 Currently _picberry_ has support for the following host boards/processors:

- the [Raspberry Pi](https://www.raspberrypi.org/)
- Allwinner A10-based boards (like the [Cubieboard](http://cubieboard.org/))
- TI AM335x-based boards (like the [Beaglebone Black](https://beagleboard.org/black) or the [AM3359 ICEv2](http://www.ti.com/tool/tmdsice3359)).

Support for additional boards and processors can be easily added, providing the following macro in a header file inside the _hosts_ folder:

	/* GPIO registers address */
	#define GPIO_BASE		// base address for the GPIO controller
	#define BLOCK_SIZE		// size of the GPIO bank
	#define PORTOFFSET		// port offset for letter-defined gpios

	/* GPIO setup macros */
	#define GPIO_IN(g)		// set gpio g as input
	#define GPIO_OUT(g)		// set gpio g as output
	#define GPIO_SET(g)		// set gpio g as high
	#define GPIO_CLR(g)		// set gpio g as low
	#define GPIO_LEV(g)		// read level of gpio g

	/* default GPIO <-> PIC connections */
	#define DEFAULT_PIC_CLK		// default gpio for PGC line
	#define DEFAULT_PIC_DATA	// default gpio for PGD line
	#define DEFAULT_PIC_MCLR	// default gpio for MCLR line

A build rule inside the Makefile for the specific platform has to be added too.

## Building and Installing picberry

picberry is written using C++11 features, so it requires at least g++ version 4.7.

To build picberry launch `make TARGET`, where _TARGET_ can be one of the following:

|   _TARGET_    | Processor/Board                            |
| ------------- | ------------------------------------------ |
| raspberrypi   | Raspberry Pi v1 or Zero                    |
| raspberrypi2  | Raspberry Pi v2 or v3                      |
| am335x        | Boards based on TI AM335x (BeagleBone)     |
| a10           | Boards based on Allwinner A10 (Cubieboard) |

Then launch `sudo make install` to install it to /usr/bin.

To change destination prefix use PREFIX=, e.g. `sudo make install PREFIX=/usr/local`.

For cross-compilation, given that you have the required cross toolchain in you PATH, simply export the `CROSS_COMPILE` variable before launching `make`, e.g. `CROSS_COMPILE=arm-linux-gnueabihf- make raspberrypi2`.

## Using picberry

	picberry [options]

Programming Options

	-h                print help
	-g PGC,PGD,MCLR   GPIO selection in form PORT:NUM, default if not present
	-S port			  server mode, listening on given port                
	-f family		  PIC family (dspic33e, dspic33f, pic18fj, pic32mx1, pic32mx2, pic32mx3, pic32mz)
	-i file           input file
	-o file           output file (default: ofile.hex)

	-r                read chip
	-w                bulk erase and write chip
	-e                bulk erase chip
	-b                blank check of the chip
	-d                read configuration registers

    --noverify        skip memory verification after programming
    --debug           turn ON debug

Runtime Options

       -R                reset

For Example, to connect the PIC to RPi GPIOs 11 (PGC), 9 (PGD), 22 (MCLR) and write on a dsPIC33FJ128GP802 the file fw.hex:

	picberry -w -g 11,9,22 -f dspic33f -i fw.hex

To connect the PIC to A10 GPIOs B15 (PGC), B17 (PGD), I15 (MCLR):

	picberry -w -g B:15,B:17,I:15 -f dspic33f -i fw.hex

### Programming Hardware

To use picberry you will need only the "recommended minimum connections" outlined in each PIC datasheet.

Between PIC and the SoC you must have the four basic ICSP lines: PGC (clock), PGD (data), MCLR (Reset), GND.
You can also connect PIC VDD line to target board 3v3 line, but be careful: such pins normally have low current capabilities, so consider your circuit current drawn!

If not specified in the command line, the default GPIOs <-> PIC connections are the following:

for the Raspberry Pi:

	PGC  <-> GPIO23
	PGD  <-> GPIO24
	MCLR <-> GPIO18

for the Allwinner A10:

	PGC  <-> PB15
	PGD  <-> PB17
	MCLR <-> PB12

for AM335x:

	PGC  <-> GPIO12
	PGD  <-> GPIO13
	MCLR <-> GPIO40


### Remote GUI

Remote GUI is written in Qt5 and allows to control a picberry session running in *server mode* (that is, launched with the  `-S <port>` command line argument).

To compile it, just launch `qmake` and then `make` in the *remote_gui* folder.

## References

- [dsPIC33E/PIC24E Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/70619B.pdf)
- [dsPIC33F/PIC24H Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/70152H.pdf)
- [PIC24FJ Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/30010057d.pdf)
- [PIC24FJXXGA0XX Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/39768d.pdf)
- [PIC24FJXXXDA1/DA2/GB2/GA3/GC0 Families Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/39970e.pdf)
- [PIC10(L)F320/322 Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/41572D.pdf)
- [PIC12(L)F1822/PIC16(L)F182X Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/41390D.pdf)
- [PIC18F2XJXX/4XJXX Family Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/39687e.pdf)
- [PIC32 Flash Programming Specification](http://ww1.microchip.com/downloads/en/DeviceDoc/60001145S.pdf)

## Licensing

picberry is released under the GPLv3 license; for full license see the `LICENSE` file.

The Microchip name and logo, PIC, In-Circuit Serial Programming, ICSP are registered trademarks of Microchip Technology Incorporated in the U.S.A. and other countries.

Raspberry Pi is a trademark of The Raspberry Pi Foundation.


debian/changelog
picberry (2.0.2) unstable; urgency=low

  * Initial release but second version.

 -- Ulysse VAUTIER <ulysse.vautier@gmail.com>  Mon, 01 Oct 2018 18:34:53 +0100


src/common.h
/*
 * Raspberry Pi PIC Programmer using GPIO connector
 * https://github.com/WallaceIT/picberry
 * Copyright 2014 Francesco Valla
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef COMMON_H_
#define COMMON_H_

#if defined(BOARD_A10)
#include "hosts/a10.h"
#elif defined(BOARD_RPI)
#include "hosts/rpi.h"
#elif defined(BOARD_RPI2)
#include "hosts/rpi2.h"
#elif defined(BOARD_AM335X)
#include "hosts/am335x.h"
#endif

#include "devices/device.h"

using namespace std;

#define VERSION "1.0.r"

/* Low-level functions */
void delay_us(unsigned int howLong);
void setup_io(void);
void close_io(void);

/* inhx.cpp functions */
unsigned int read_inhx(char *infile, memory *mem, uint32_t offset=0);
void write_inhx(memory *mem, char *outfile, uint32_t offset=0);

/* Runtime Functions */
void pic_reset(bool silent = false);

/* main functions */
void usage(void);
void server_mode(int port);
uint8_t send_file(char * filename);
uint8_t receive_file(int sock, char * filename);

extern volatile uint32_t *gpio;
extern int pic_clk, pic_data, pic_mclr;

struct flags_struct {
   int debug = 0;
   int client = 0;
   int noverify = 0;
   int boot_only = 0;
   int program_only = 0;
};

struct detailed_subfamily_t{
	uint32_t device_id;
	uint8_t detailed_subfamily;
	uint8_t latch_size;
};

extern struct flags_struct flags;

#endif /* COMMON_H_ */


src/devices/pic10f322.h
/*
 * Raspberry Pi PIC Programmer using GPIO connector
 * https://github.com/WallaceIT/picberry
 * Copyright 2014 Francesco Valla
 * Copyright 2017 Akimasa Tateba
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <iostream>

#include "../common.h"
#include "device.h"

using namespace std;

#define SF_PIC10F322		0x00
#define SF_PIC12F1822		0x01
#define SF_PIC16LF1826		0x02

class pic10f322: public Pic{

	public:
		uint8_t detailed_subfamily;
		uint8_t	latch_size;
		pic10f322(void){
			latch_size = 0;
		};
		void enter_program_mode(void);
		void exit_program_mode(void);
		bool setup_pe(void){return true;};
		bool read_device_id(void);
		void bulk_erase(void);
		void dump_configuration_registers(void);
		void read(char *outfile, uint32_t start, uint32_t count);
		void write(char *infile);
		uint8_t blank_check(void);

	protected:
		void send_cmd(uint8_t cmd, unsigned int delay);
		uint16_t read_data(void);
		void write_data(uint16_t data);
		void reset_mem_location(void);

		/*
		* DEVICES SECTION
		*                    	ID       NAME           MEMSIZE
		*/
		pic_device piclist[19] = {{0x14D,  "PIC10F320", 0x100},
								{0x14C,  "PIC10F322", 0x200},
								{0x14F,  "PIC10LF320", 0x100},
								{0x13C,  "PIC16F1826", 0x800},
								{0x13D,  "PIC16F1827", 0x1000},
								{0x144,  "PIC16LF1826", 0x800},
								{0x145,  "PIC16LF1827", 0x1000},
								{0x139,  "PIC16F1823", 0x800},
								{0x141,  "PICLF1823", 0x800},
								{0x138,  "PIC12F1822", 0x800},
								{0x140,  "PIC12LF1822", 0x800},
								{0x13A,  "PIC16F1824", 0x1000},
								{0x142,  "PIC16LF1824", 0x1000},
								{0x13B,  "PIC16F1825", 0x2000},
								{0x143,  "PIC16LF1825", 0x2000},
								{0x13E,  "PIC16F1828", 0x1000},
								{0x146,  "PIC16LF1828", 0x1000},
								{0x13F,  "PIC16F1829", 0x2000},
								{0x147,  "PIC16LF1829", 0x2000}
								};
		detailed_subfamily_t detailed_subfamily_table[19] = {
								{0x14D,SF_PIC10F322,	16},	//PIC10F320
								{0x14C,SF_PIC10F322,	16},	//PIC10F322
								{0x14F,SF_PIC10F322,	16},	//PIC10LF320
								{0x13C,SF_PIC12F1822,	8},	//PIC16F1826
								{0x13D,SF_PIC12F1822,	8},	//PIC16F1827
								{0x144,SF_PIC16LF1826,	8},	//PIC16LF1826
								{0x145,SF_PIC16LF1826,	8},	//PIC16LF1827
								{0x139,SF_PIC12F1822,	16},	//PIC16F1823
								{0x141,SF_PIC12F1822,	16},	//PICLF1823
								{0x138,SF_PIC12F1822,	16},	//PIC12F1822
								{0x140,SF_PIC12F1822,	16},	//PIC12LF1822
								{0x13A,SF_PIC12F1822,	32},	//PIC16F1824
								{0x142,SF_PIC12F1822,	32},	//PIC16LF1824
								{0x13B,SF_PIC12F1822,	32},	//PIC16F1825
								{0x143,SF_PIC12F1822,	32},	//PIC16LF1825
								{0x13E,SF_PIC12F1822,	32},	//PIC16F1828
								{0x146,SF_PIC12F1822,	32},	//PIC16LF1828
								{0x13F,SF_PIC12F1822,	32},	//PIC16F1829
								{0x147,SF_PIC12F1822,	32}	//PIC16LF1829
								};
};


src/devices/pic16f150x.cpp
/*
 * Raspberry Pi PIC Programmer using GPIO connector
 * https://github.com/WallaceIT/picberry
 * Copyright 2014 Francesco Valla
 * Copyright 2017 Akimasa Tateba
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * Author : Ulysse VAUTIER
 *
 */

#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <iostream>

#include "pic16f150x.h"

/* delays (in microseconds) */
#define DELAY_SETUP	1
#define	DELAY_HOLD	1
#define DELAY_TENTS	1
#define DELAY_TENTH	250
#define DELAY_TCKH	1
#define DELAY_TCKL 	1
#define DELAY_TCO 	1
#define DELAY_TDLY	1
#define DELAY_TERAB	5000
#define DELAY_TEXIT	1
#define DELAY_TPINT_DATA	2500
#define DELAY_TPINT_CONF	5000

/* commands for programming */
#define COMM_LOAD_CONFIG	0x00
#define COMM_LOAD_FOR_PROG	0x02
#define COMM_READ_FROM_PROG	0x04
#define COMM_INC_ADDR		0x06
#define COMM_RESET_ADDR		0x16
#define COMM_BEGIN_IN_TIMED_PROG	0x08
#define COMM_BEGIN_EXT_TIMED_PROG	0x18
#define COMM_END_EXT_TIMED_PROG	0x0A
#define COMM_BULK_ERASE		0x09
#define COMM_ROW_ERASE		0x11

#define ENTER_PROGRAM_KEY	0x4D434850

void pic16f150x::enter_program_mode(void)
{
	int i;

	GPIO_IN(pic_mclr);
	GPIO_OUT(pic_mclr);

	GPIO_SET(pic_mclr);			/* apply VDD to MCLR pin */
	delay_us(DELAY_TENTS);	/* wait TENTS */
	GPIO_CLR(pic_mclr);			/* remove VDD from MCLR pin */
	GPIO_CLR(pic_clk);
	delay_us(DELAY_TENTH);		/* wait TENTH */
	/* Shift in the "enter program mode" key sequence (LSB! first) */
	for (i = 0; i < 32; i++) {
		if ( (ENTER_PROGRAM_KEY >> i) & 0x01 )
			GPIO_SET(pic_data);
		else
			GPIO_CLR(pic_data);

		delay_us(DELAY_TCKL);	/* Setup time */
		GPIO_SET(pic_clk);
		delay_us(DELAY_TCKH);	/* Hold time */
		GPIO_CLR(pic_clk);

	}
	GPIO_CLR(pic_data);

	//Last clock(Don't care data)
	delay_us(DELAY_TCKL);	/* Setup time */
	GPIO_SET(pic_clk);
	delay_us(DELAY_TCKH);	/* Hold time */
	GPIO_CLR(pic_clk);

}

void pic16f150x::exit_program_mode(void)
{

	GPIO_CLR(pic_clk);			/* stop clock on PGC */
	GPIO_CLR(pic_data);			/* clear data pin PGD */

	GPIO_IN(pic_mclr);
}

/* Send a 4-bit command to the PIC (LSB first) */
void pic16f150x::send_cmd(uint8_t cmd, unsigned int delay)
{
	int i;

	for (i = 0; i < 6; i++) {
		GPIO_SET(pic_clk);
		if ( (cmd >> i) & 0x01 )
			GPIO_SET(pic_data);
		else
			GPIO_CLR(pic_data);
		delay_us(DELAY_TCKH);	/* Setup time */
		GPIO_CLR(pic_clk);
		delay_us(DELAY_TCKL);	/* Hold time */
	}
	GPIO_CLR(pic_data);
	delay_us(delay);
}

/* Read 8-bit data from the PIC (LSB first) */
uint16_t pic16f150x::read_data(void)
{
	uint8_t i;
	uint16_t data = 0x0000;

	GPIO_IN(pic_data);

	for (i = 0; i < 16; i++) {
		GPIO_SET(pic_clk);
		delay_us(DELAY_TCKH);
		delay_us(DELAY_TCO);	/* Wait for data to be valid */
		data |= ( GPIO_LEV(pic_data) & 0x00000001 ) << i;
		GPIO_CLR(pic_clk);
		delay_us(DELAY_TCKL);
	}

	GPIO_IN(pic_data);
	GPIO_OUT(pic_data);
	data >>= 1;
	return data;
}

/* Load 16-bit data to the PIC (LSB first) */
void pic16f150x::write_data(uint16_t data)
{
	int i;
	data <<= 1;

	for (i = 0; i < 16; i++) {
		GPIO_SET(pic_clk);
		if ( (data >> i) & 0x0001 )
			GPIO_SET(pic_data);
		else
			GPIO_CLR(pic_data);
		delay_us(DELAY_SETUP);	/* Setup time */
		GPIO_CLR(pic_clk);
		delay_us(DELAY_HOLD);	/* Hold time */
	}
	GPIO_CLR(pic_data);
}

/* set Table Pointer */
void pic16f150x::reset_mem_location(void)
{
	send_cmd(COMM_RESET_ADDR, DELAY_TDLY);
}

/* Read PIC device id word */
bool pic16f150x::read_device_id(void)
{
	uint16_t id;
	bool found = 0, found2 = 0;

	send_cmd(COMM_LOAD_CONFIG, DELAY_TDLY);
	write_data(0x00);

	for(int i=0; i < 6; i++){
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);
	}
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);
	id = read_data();
	device_id = (id >> 5) & 0x1ff;
	device_rev = id & 0x1f;

	for (unsigned short i=0;i < sizeof(piclist)/sizeof(piclist[0]);i++){

		if (piclist[i].device_id == device_id){

			strcpy(name,piclist[i].name);
			mem.code_memory_size = piclist[i].code_memory_size;
			mem.program_memory_size = 0x0F80018;
			mem.location = (uint16_t*) calloc(mem.program_memory_size,sizeof(uint16_t));
			mem.filled = (bool*) calloc(mem.program_memory_size,sizeof(bool));
			found = 1;
			break;
		}
	}
	for (unsigned short i=0;i < sizeof(detailed_subfamily_table)/sizeof(detailed_subfamily_table[0]);i++){
		if (detailed_subfamily_table[i].device_id == device_id){
			detailed_subfamily = detailed_subfamily_table[i].detailed_subfamily;
			latch_size = detailed_subfamily_table[i].latch_size;			
			found2 = 1;
			break;
		}
	}
	return found & found2;

}

/* Blank Check */
uint8_t pic16f150x::blank_check(void)
{
	unsigned int lcounter = 0;

	uint16_t addr, data;
	uint8_t ret = 0;

	if(!flags.debug) cerr << "[ 0%]";
	lcounter = 0;

	reset_mem_location();

	for(addr = 0; addr < mem.code_memory_size; addr++){
		send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);
		data = read_data() & 0x3FFF;
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);

		if(data != 0x3FFF) {
			fprintf(stderr, "Chip not Blank! Address: 0x%x, Read: 0x%x.\n",  addr, data);
			ret = 1;
			break;
		}

		if(lcounter != addr*100/mem.code_memory_size){
			lcounter = addr*100/mem.code_memory_size;
			fprintf(stderr, "\b\b\b\b\b[%2d%%]", lcounter);
		}
	}

	/* Read Confuguration Fuses */
	send_cmd(COMM_LOAD_CONFIG, DELAY_TDLY);
	write_data(0x00);

	addr = 0x8000;
	for(int i = 0; i < 7; i++){
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);
		addr++;
	}
	/* Config Word 1 */
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);

	data = read_data() & 0x3FFF;

	if (data != 0x3FFF) {
		fprintf(stderr, "Chip not Blank! Address: 0x%x, Read: 0x%x.\n",  addr, data);
		ret = 1;
	}
	/* Config Word 2 */
	addr++;
	send_cmd(COMM_INC_ADDR, DELAY_TDLY);
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);

	data = read_data() & 0x3FFF;

	if (data != 0x3FFF) {
		fprintf(stderr, "Chip not Blank! Address: 0x%x, Read: 0x%x.\n",  addr, data);
		ret = 1;
	}

	if(!flags.debug) cerr << "\b\b\b\b\b";

	return ret;

}

/* Bulk erase the chip */
void pic16f150x::bulk_erase(void)
{
	send_cmd(COMM_RESET_ADDR, DELAY_TDLY);
	send_cmd(COMM_BULK_ERASE, DELAY_TERAB);
	if(flags.client) fprintf(stdout, "@FIN");
}

/* Read PIC memory and write the contents to a .hex file */
void pic16f150x::read(char *outfile, uint32_t start, uint32_t count)
{
	uint16_t addr, data = 0x0000;

	if(!flags.debug) cerr << "[ 0%]";
	if(flags.client) fprintf(stdout, "@000");
	unsigned int lcounter = 0;

	/* Read Memory */

	reset_mem_location();

	for (addr = 0; addr < mem.code_memory_size; addr++) {
		send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);
		data = read_data() & 0x3FFF;
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);

		if (flags.debug)
			fprintf(stderr, "  addr = 0x%04X  data = 0x%04X\n", addr, data);

		if (data != 0x3FFF) {
			mem.location[addr]        = data;
			mem.filled[addr]      = 1;
		}

		if(lcounter != addr*100/mem.code_memory_size){
			if(flags.client)
				fprintf(stderr,"RED@%2d\n", (addr*100/mem.code_memory_size));
			if(!flags.debug)
				fprintf(stderr,"\b\b\b\b%2d%%]", addr*100/mem.code_memory_size);
			lcounter = addr*100/mem.code_memory_size;
		}
	}
	/* Read Confuguration Fuses */
	send_cmd(COMM_LOAD_CONFIG, DELAY_TDLY);
	write_data(0x00);

	addr = 0x8000;
	for(int i = 0; i < 7; i++){
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);
		addr++;
	}
	/* Config Word 1 */
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);

	data = read_data() & 0x3FFF;

	if (flags.debug)
		fprintf(stderr, "  addr = 0x%04X  data = 0x%04X\n", addr, data);

	if (data != 0x3FFF) {
		mem.location[addr]        = data;
		mem.filled[addr]      = 1;
	}
	/* Config Word 2 */
	uint16_t mask = 0x3E03;

	addr++;
	send_cmd(COMM_INC_ADDR, DELAY_TDLY);
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);

	if(detailed_subfamily == SF_PIC16F1501_7)
		mask = 0x2E03;

	data = read_data() & mask;

	if (flags.debug)
		fprintf(stderr, "  addr = 0x%04X  data = 0x%04X\n", addr, data);

	if (data != mask) {
		mem.location[addr]        = data;
		mem.filled[addr]      = 1;
	}

	if(!flags.debug) cerr << "\b\b\b\b\b";
	if(flags.client) fprintf(stdout, "@FIN");
	write_inhx(&mem, outfile);
}

/* Bulk erase the chip, and then write contents of the .hex file to the PIC */
void pic16f150x::write(char *infile)
{
	int i;
	uint16_t data, fileconf;
	uint32_t addr = 0x00000000;

	read_inhx(infile, &mem);

	bulk_erase();

	if(!flags.debug) cerr << "[ 0%]";
	if(flags.client) fprintf(stdout, "@000");
	unsigned int lcounter = 0;

	reset_mem_location();

	for (addr = 0; addr < mem.code_memory_size; addr += latch_size){        /* address in WORDS (2 Bytes) */


		if (flags.debug)
			fprintf(stderr, "Current address 0x%08X \n", addr);
		for(i=0; i<latch_size-1; i++){		                        /* write the first 62 bytes */
			if (mem.filled[addr+i]) {
				if (flags.debug)
					fprintf(stderr, "  Writing 0x%04X to address 0x%06X \n", mem.location[addr + i], (addr+i) );
				send_cmd(COMM_LOAD_FOR_PROG, DELAY_TDLY);
				write_data(mem.location[addr+i]);
			}
			else {
				if (flags.debug)
					fprintf(stderr, "  Writing 0x3FFF to address 0x%06X \n", (addr+i) );
				send_cmd(COMM_LOAD_FOR_PROG, DELAY_TDLY);
				write_data(0x3FFF);			/* write 0x3FFF in empty locations */
			};
			send_cmd(COMM_INC_ADDR, DELAY_TDLY);
		}

		/* write the last 2 bytes and start programming */
		if (mem.filled[addr+latch_size-1]) {
			if (flags.debug)
				fprintf(stderr, "  Writing 0x%04X to address 0x%06X and then start programming...\n", mem.location[addr+latch_size-1], (addr+latch_size-1));
			send_cmd(COMM_LOAD_FOR_PROG, DELAY_TDLY);
			write_data(mem.location[addr+latch_size-1]);
		}
		else {
			if (flags.debug)
				fprintf(stderr, "  Writing 0x3FFF to address 0x%06X and then start programming...\n", (addr+latch_size-1));
			send_cmd(COMM_LOAD_FOR_PROG, DELAY_TDLY);
			write_data(0x3FFF);			         /* write 0x3FFF in empty locations */
		};

		/* Programming Sequence */
		send_cmd(COMM_BEGIN_IN_TIMED_PROG, DELAY_TPINT_DATA);
		/* end of Programming Sequence */

		send_cmd(COMM_INC_ADDR, DELAY_TDLY);

		if(lcounter != addr*100/mem.code_memory_size){
			lcounter = addr*100/mem.code_memory_size;
			if(flags.client)
				fprintf(stdout,"@%03d", lcounter);
			if(!flags.debug)
				fprintf(stderr,"\b\b\b\b\b[%2d%%]", lcounter);
		}
	};

	if(!flags.debug) cerr << "\b\b\b\b\b\b";
	if(flags.client) fprintf(stdout, "@100");

	/* Write Confuguration Fuses
	 * Writing User ID is not implemented.
	 */
	send_cmd(COMM_LOAD_CONFIG, DELAY_TDLY);
	write_data(0x00);

	addr = 0x8000;
	for(i = 0; i < 7; i++){
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);
		addr++;
	}
	if(mem.filled[addr]){
		send_cmd(COMM_LOAD_FOR_PROG, DELAY_TDLY);
		write_data(mem.location[addr]);

		send_cmd(COMM_BEGIN_IN_TIMED_PROG, DELAY_TPINT_CONF);
	}

	addr++;
	send_cmd(COMM_INC_ADDR, DELAY_TDLY);
	if(mem.filled[addr]){
		send_cmd(COMM_LOAD_FOR_PROG, DELAY_TDLY);
		write_data(mem.location[addr]);

		send_cmd(COMM_BEGIN_IN_TIMED_PROG, DELAY_TPINT_CONF);
	}
	/* Verify Code Memory and Configuration Word */
	if(!flags.noverify){
		if(!flags.debug) cerr << "[ 0%]";
		if(flags.client) fprintf(stdout, "@000");
		lcounter = 0;

		reset_mem_location();

		for (addr = 0; addr < mem.code_memory_size; addr++) {
			send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);
			data = read_data() & 0x3FFF;
			send_cmd(COMM_INC_ADDR, DELAY_TDLY);

			if (flags.debug)
				fprintf(stderr, "addr = 0x%06X:  pic = 0x%04X, file = 0x%04X\n",
						addr, data, (mem.filled[addr]) ? (mem.location[addr]) : 0x3FFF);

			if ( (data != mem.location[addr]) & ( mem.filled[addr]) ) {
				fprintf(stderr, "Error at addr = 0x%06X:  pic = 0x%04X, file = 0x%04X.\nExiting...",
						addr, data, mem.location[addr]);
				return;
			}
			if(lcounter != addr*100/mem.code_memory_size){
				lcounter = addr*100/mem.code_memory_size;
				if(flags.client)
					fprintf(stdout,"@%03d", lcounter);
				if(!flags.debug)
					fprintf(stderr,"\b\b\b\b\b[%2d%%]", lcounter);
			}
		}

		/* Read Confuguration Fuses */
		send_cmd(COMM_LOAD_CONFIG, DELAY_TDLY);
		write_data(0x00);

		addr = 0x8000;
		for(int i = 0; i < 7; i++){
			send_cmd(COMM_INC_ADDR, DELAY_TDLY);
			addr++;
		}

		send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);

		/* NOTE: It is impossible to program LVP bit when Low-Voltage Programming.
		 * We will ignore LVP bit in Configuration Fuse by using 0x3E03 mask.
		 */
		uint16_t mask = 0x3FFF;
		//if(detailed_subfamily == SF_PIC16F1501_7)
		//	mask = 0x2E03;

		data = read_data() & mask;
		fileconf = mem.location[addr] & mask;
		if ( ( data != fileconf ) & ( mem.filled[addr] ) ) {
			fprintf(stderr, "Error at addr = 0x%06X:  pic = 0x%04X, file = 0x%04X.\nExiting...",
					addr, data, mem.location[addr] & mask);
			return;
		}

		/* Config Word 2 */
		mask = 0x3E03;
		addr++;
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);
		send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);

		if(detailed_subfamily == SF_PIC16F1501_7)
			mask = 0x2E03;

		/* Ignore LVP bit. */
		mask &= ~(1 << 13);
		data = read_data() & mask;
		fileconf = mem.location[addr] & mask;
		if ( ( data != fileconf ) & ( mem.filled[addr] ) ) {
			fprintf(stderr, "Error at addr = 0x%06X:  pic = 0x%04X, file = 0x%04X.\nExiting...",
					addr, data & mask, mem.location[addr] & mask);
			return;
		}
	}

	if(!flags.debug) cerr << "\b\b\b\b\b";
	if(flags.client) fprintf(stdout, "@FIN");
	else{
		if(flags.client) fprintf(stdout, "@FIN");
	}

}

/* Dum configuration words */
void pic16f150x::dump_configuration_registers(void)
{
	send_cmd(COMM_LOAD_CONFIG, DELAY_TDLY);
	write_data(0x00);

	for(int i=0; i < 7; i++){
		send_cmd(COMM_INC_ADDR, DELAY_TDLY);
	}
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);
	cout << "Configuration Words:" << endl;
	fprintf(stdout, " - CONFIG1 = 0x%2x.\n", (read_data() & 0x3FFF));
	send_cmd(COMM_INC_ADDR, DELAY_TDLY);
	send_cmd(COMM_READ_FROM_PROG, DELAY_TDLY);
	fprintf(stdout, " - CONFIG2 = 0x%2x.\n", (read_data() & 0x3FFF));
}
